File: ios/AbrevvaCrypto.swift
import AbrevvaSDK
import CryptoKit
import CryptoSwift
import Foundation

private enum CryptoError: Error {
  case EncryptCryptoError
  case DecryptCryptoError
  case ComputeSharedSecretError
  case DecryptFileReadError
  case DecryptFileFromURLNetworkError
  case DecryptFileFromURLNotFoundError
  case DecryptFileFromURLInaccessibleError
  case DecryptFileFromURLNoResponseDataError
  case ComputeED25519PublicKeyError
  case SignCryptoError
}

class AbrevvaCryptoImpl: HybridAbrevvaCryptoImplSpec {
  private let X25519Impl = X25519()
  private let AesGcmImpl = AesGcm()
  private let AesCcmImpl = AesCcm()
  private let SimpleSecureRandomImpl = SimpleSecureRandom()
  private let HKDFImpl = HKDFWrapper()
  
  func encrypt(key: String, iv: String, adata: String, pt: String, tagLength: Double?) throws -> EncryptResult {
    let ct = AesCcmImpl.encrypt(
      key: [UInt8](hex: "0x" + key),
      iv: [UInt8](hex: "0x" + iv),
      adata: [UInt8](hex: "0x" + adata),
      pt: [UInt8](hex: "0x" + pt),
      tagLength: Int(tagLength ?? 0)
    )
    if ct.isEmpty {
      throw CryptoError.EncryptCryptoError
    } else {
      return EncryptResult(
        cipherText: [UInt8](ct[..<pt.count]).toHexString(),
        authTag: [UInt8](ct[pt.count...]).toHexString()
      )
    }
  }
  
  func decrypt(key: String, iv: String, adata: String, ct: String, tagLength: Double?) throws -> DecryptResult {
    
    let pt = AesCcmImpl.decrypt(
      key: [UInt8](hex: "0x" + key),
      iv: [UInt8](hex: "0x" + iv),
      adata: [UInt8](hex: "0x" + adata),
      ct: [UInt8](hex: "0x" + ct),
      tagLength: Int(tagLength ?? 0)
    ).toHexString()
    if pt.isEmpty {
      throw CryptoError.DecryptCryptoError
    } else {
      return DecryptResult(plainText: pt, authOk: true)
    }
  }
  
  func generateKeyPair() -> KeypairResult {
    let keyPair = X25519Impl.generateKeyPair()
    
    return KeypairResult(
      privateKey: keyPair[0].toHexString(),
      publicKey: keyPair[1].toHexString()
    )
  }
  
  func computeSharedSecret(key: String, peerPublicKey: String) throws -> String {
    let privateKeyData = Data(hex: "0x" + key)
    let publicKeyData = Data(hex: "0x" + peerPublicKey)
    let sharedSecret = X25519Impl.computeSharedSecret(
      privateKeyData: privateKeyData, publicKeyData: publicKeyData
    )
    if sharedSecret == nil {
      throw CryptoError.ComputeSharedSecretError
    }
    return sharedSecret!.toHexString()
  }
  
  func encryptFile(sharedSecret: String, ptPath: String, ctPath: String) -> Bool {
    let sharedSecret = [UInt8](
      hex: "0x" + sharedSecret)
    
    let operationResult = AesGcmImpl.encryptFile(
      key: sharedSecret, pathPt: ptPath, pathCt: ctPath
    )
    return operationResult
  }
  
  func decryptFile(sharedSecret: String, ptPath: String, ctPath: String) throws -> Bool {
    let sharedSecret = [UInt8](
      hex: "0x" + sharedSecret)
    
    let url = URL(fileURLWithPath: ctPath)
    
    let data: Data
    do {
      data = try Data(contentsOf: url, options: .mappedIfSafe)
    } catch {
      throw CryptoError.DecryptFileReadError
    }
    
    return AesGcmImpl.decryptFile(
      key: sharedSecret, data: data, pathPt: ptPath
    )
  }
  
    func decryptFileFromURL(sharedSecret: String, ptPath: String, url: String) throws -> NitroModules.Promise<Bool> {
        let sharedSecret = [UInt8](
            hex: "0x" + sharedSecret)
        var result: (Data, URLResponse)?
        var e: Error?
        let semaphore = DispatchSemaphore(value: 0)
        
        let promise = Promise<Bool>()
        Task {
            do {
                guard let url = URL(string: url) else {
                    throw CryptoError.DecryptFileFromURLNetworkError
                }
                result = try await URLSession.shared.data(from: url)
            } catch {
                e = error
            }
            semaphore.signal()
        }
        semaphore.wait()
        
        if e != nil {
            return Promise.rejected(withError: CryptoError.DecryptFileFromURLNetworkError)
        }
        
        guard  let httpResult = result, let httpResponse = httpResult.1 as? HTTPURLResponse else {
            return Promise.rejected(withError: CryptoError.DecryptFileFromURLNoResponseDataError)
        }
        
        if httpResponse.statusCode == 200 {
            promise.resolve(withResult: self.AesGcmImpl.decryptFile(key: sharedSecret, data: httpResult.0, pathPt: ptPath))
        } else if  httpResponse.statusCode == 404 {
            return Promise.rejected(withError: CryptoError.DecryptFileFromURLNotFoundError)
        }
        return Promise.rejected(withError: CryptoError.DecryptFileFromURLInaccessibleError)
    }
  
  func random(numBytes: Double) -> String {
    return SimpleSecureRandomImpl.random(Int(numBytes)).toHexString()
  }
  
  func derive(key: String, salt: String, info: String, length: Double) -> String {
    return HKDFImpl.derive(
      key: [UInt8](hex: "0x" + key),
      salt: [UInt8](hex: "0x" + salt),
      info: [UInt8](hex: "0x" + info),
      length: Int(length)
    ).toHexString()
  }
  
  func computeED25519PublicKey(privateKey: String) throws -> String {
    let privateKeyData = Data(base64Encoded: privateKey)
    
    if let publicKey = self.X25519Impl.computeED25519PublicKey(privateKeyData: privateKeyData!) {
      return publicKey.base64EncodedString()
    }
    throw CryptoError.ComputeED25519PublicKeyError
  }
  
  func sign(privateKey: String, data: String) throws -> String {
    let privateKeyData = Data(base64Encoded: privateKey)
    let data = data.data(using: .utf8)
    
    if let signature = self.X25519Impl.sign(privateKeyData: privateKeyData!, data: data!) {
      return signature.base64EncodedString()
    }
    throw CryptoError.SignCryptoError
  }
  
  func verify(publicKey: String, data: String, signature: String) -> Bool {
    let publicKeyData = Data(base64Encoded: publicKey)
    let data = data.data(using: .utf8)
    let signature = Data(base64Encoded: signature)
    
    return self.X25519Impl.verify(publicKeyData: publicKeyData!, data: data!, signature: signature!)
  }
}

File: android/src/main/java/com/margelo/nitro/abrevvareactnative/AbrevvaCrypto.kt
package com.margelo.nitro.abrevvareactnative

import com.evva.xesar.abrevva.crypto.AesCcm
import com.evva.xesar.abrevva.crypto.AesGcm
import com.evva.xesar.abrevva.crypto.HKDF
import com.evva.xesar.abrevva.crypto.SimpleSecureRandom
import com.evva.xesar.abrevva.crypto.X25519Wrapper
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.bouncycastle.util.encoders.Base64
import org.bouncycastle.util.encoders.Hex
import java.io.BufferedInputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.nio.file.Paths

private enum class CryptoError {
  EncryptCryptoError,
  EncryptEmptyResultError,
  EncryptFileCryptoError,
  EncryptFileInvalidArgumentError,
  DecryptEmptyResultError,
  DecryptCryptoError,
  DecryptFileCryptoError,
  DecryptFileInvalidArgumentError,
  DecryptFileFromURLNetworkError,
  DecryptFileFromURLNotFoundError,
  DecryptFileFromURLInaccessibleError,
  DecryptFileFromURLNoResponseDataError,
  DecryptFileFromURLInvalidArgumentError,
  GenerateKeypairError,
  ComputeSharedSecretError,
  ComputeED25519PublicKeyError,
  SignCryptoError,
  VerifyCryptoError,
  RandomError,
  DeriveInvalidArgumentError,
  DeriveEmptyResultError,
  DeriveCryptoError
}

private class CryptoException(error: CryptoError, throwable: Throwable? = null) :
  Exception("${AbrevvaCryptoImpl::class.simpleName}::${error.name}", throwable)

@DoNotStrip
class AbrevvaCryptoImpl : HybridAbrevvaCryptoImplSpec() {
  override fun encrypt(
    key: String,
    iv: String,
    adata: String,
    pt: String,
    tagLength: Double?
  ): EncryptResult {
    val keyByte = Hex.decode(key)
    val ivByte = Hex.decode(iv)
    val adataByte = Hex.decode(adata)
    val ptByte = Hex.decode(pt)
    val _tagLength: Int = (tagLength ?: DEFAULT_TIMEOUT).toInt()

    try {
      val ct: ByteArray = AesCcm.encrypt(keyByte, ivByte, adataByte, ptByte, _tagLength)
      val cipherText = ByteArray(ptByte.size)
      val authTag = ByteArray(_tagLength)

      System.arraycopy(ct, 0, cipherText, 0, ptByte.size)
      System.arraycopy(ct, ptByte.size, authTag, 0, _tagLength)

      if (ct.isEmpty()) {
        throw CryptoException(CryptoError.EncryptEmptyResultError)
      } else {
        return EncryptResult(Hex.toHexString(cipherText), Hex.toHexString(authTag))
      }
    } catch (e: Exception) {
      throw CryptoException(CryptoError.EncryptCryptoError, e)
    }
  }

  override fun decrypt(
    key: String,
    iv: String,
    adata: String,
    ct: String,
    tagLength: Double?
  ): DecryptResult {
    val keyHex = Hex.decode(key)
    val ivHex = Hex.decode(iv)
    val adataHex = Hex.decode(adata)
    val ctHex = Hex.decode(ct)
    val _tagLength: Int = (tagLength ?: DEFAULT_TIMEOUT).toInt()

    try {
      val data: ByteArray = AesCcm.decrypt(keyHex, ivHex, adataHex, ctHex, _tagLength)
      if (data.isEmpty()) {
        throw CryptoException(CryptoError.DecryptEmptyResultError)
      }
      return DecryptResult(Hex.toHexString(data), true)
    } catch (e: Exception) {
      throw CryptoException(CryptoError.DecryptCryptoError, e)
    }
  }

  override fun generateKeyPair(): KeypairResult {
    val keyPair: X25519Wrapper.KeyPair
    val privateKey: String
    val publicKey: String
    try {
      keyPair = X25519Wrapper.generateKeyPair()
      privateKey = Base64.toBase64String(keyPair.privateKey)
      publicKey = Base64.toBase64String(keyPair.publicKey)
      return KeypairResult(privateKey, publicKey)
    } catch (e: Exception) {
      throw CryptoException(CryptoError.GenerateKeypairError, e)
    }
  }

  override fun computeSharedSecret(
    key: String,
    peerPublicKey: String
  ): String {
    try {
      val result = X25519Wrapper.computeSharedSecret(
        Base64.decode(key),
        Base64.decode(peerPublicKey)
      )
      return Hex.toHexString(result)
    } catch (e: Exception) {
      throw CryptoException(CryptoError.ComputeSharedSecretError, e)
    }
  }

  override fun encryptFile(
    sharedSecret: String,
    ptPath: String,
    ctPath: String
  ): Boolean {
    val sharedSecretBytes: ByteArray

    try {
      sharedSecretBytes = Hex.decode(sharedSecret)
    } catch (e: Exception) {
      throw CryptoException(CryptoError.EncryptFileInvalidArgumentError, e)
    }

    try {
      return AesGcm.encryptFile(Hex.decode(sharedSecretBytes), ptPath, ctPath)
    } catch (e: Exception) {
      throw CryptoException(CryptoError.EncryptFileCryptoError, e)
    }
  }

  override fun decryptFile(
    sharedSecret: String,
    ptPath: String,
    ctPath: String
  ): Boolean {
    val sharedSecretByte: ByteArray

    try {
      sharedSecretByte = Hex.decode(sharedSecret)
    } catch (e: Exception) {
      throw CryptoException(CryptoError.DecryptFileInvalidArgumentError, e)
    }

    try {
      return AesGcm.decryptFile(sharedSecretByte, ctPath, ptPath)
    } catch (e: Exception) {
      throw CryptoException(CryptoError.DecryptFileCryptoError, e)
    }
  }

    override fun decryptFileFromURL(
    sharedSecret: String,
    ptPath: String,
    url: String
  ): Promise<Boolean> {
        val promise = Promise<Boolean>()

        CoroutineScope(Dispatchers.IO).launch {

            val sharedSecretByte: ByteArray
            try {
                sharedSecretByte = Hex.decode(sharedSecret)
            } catch (e: Exception) {
                promise.reject(
                    CryptoException(
                        CryptoError.DecryptFileFromURLInvalidArgumentError,
                        e
                    )
                )
                return@launch
            }

            val ctPath = Paths.get(ptPath).parent.toString() + "/blob"

            val file = File(ctPath)
            val uri: URL
            val connection: HttpURLConnection
            val statusCode: Int

            try {
                uri = URL(url)
                connection = uri.openConnection() as HttpURLConnection
                statusCode = connection.responseCode
            } catch (e: Exception) {
                promise.reject(CryptoException(CryptoError.DecryptFileFromURLNetworkError, e))
                return@launch
            }
            try {
                when (statusCode) {
                    200 -> {
                        val inputStream = connection.inputStream
                        val bufferedInputStream = BufferedInputStream(inputStream)
                        val outputStream = FileOutputStream(file)
                        val dataBuffer = ByteArray(4096)
                        var bytesRead: Int

                        while (bufferedInputStream.read(dataBuffer, 0, 4096)
                                .also { bytesRead = it } != -1
                        ) {
                            outputStream.write(dataBuffer, 0, bytesRead)
                        }
                        outputStream.flush()
                        outputStream.close()
                    }

                    404 -> {
                        promise.reject(CryptoException(CryptoError.DecryptFileFromURLNotFoundError))
                        return@launch
                    }

                    else -> {
                        promise.reject(CryptoException(CryptoError.DecryptFileFromURLInaccessibleError))
                        return@launch

                    }
                }
            } catch (e: IOException) {
                promise.reject(
                    CryptoException(
                        CryptoError.DecryptFileFromURLNoResponseDataError,
                        e
                    )
                )
                return@launch
            }
            promise.resolve(AesGcm.decryptFile(sharedSecretByte, ctPath, ptPath))
        }.start()
        return promise
    }

  override fun random(numBytes: Double): String {
    try {
      val result = SimpleSecureRandom.getSecureRandomBytes(numBytes.toInt())
      return Hex.toHexString(result)
    } catch (e: Exception) {
      throw CryptoException(CryptoError.RandomError, e)
    }
  }

  override fun derive(
    key: String,
    salt: String,
    info: String,
    length: Double
  ): String {
    val keyByte: ByteArray
    val saltByte: ByteArray
    val infoByte: ByteArray
    try {
      keyByte = Hex.decode(key)
      saltByte = Hex.decode(salt)
      infoByte = Hex.decode(info)
    } catch (e: Exception) {
      throw CryptoException(CryptoError.DeriveInvalidArgumentError, e)
    }

    try {
      val result: ByteArray = HKDF.derive(keyByte, saltByte, infoByte, length.toInt())
      if (result.isEmpty()) {
        throw CryptoException(CryptoError.DeriveEmptyResultError)
      }
      return Hex.toHexString(result)
    } catch (e: Exception) {
      throw CryptoException(CryptoError.DeriveCryptoError, e)
    }
  }

  override fun computeED25519PublicKey(privateKey: String): String {
    try {
      val result = X25519Wrapper.computeED25519PublicKey(Base64.decode(privateKey))
      return Base64.toBase64String(result)
    } catch (e: Exception) {
      throw CryptoException(CryptoError.ComputeED25519PublicKeyError, e)
    }
  }

  override fun sign(privateKey: String, data: String): String {
    try {
      val result = X25519Wrapper.sign(Base64.decode(privateKey), data.toByteArray())
      return Base64.toBase64String(result)
    } catch (e: Exception) {
      throw CryptoException(CryptoError.SignCryptoError, e)
    }
  }

  override fun verify(publicKey: String, data: String, signature: String): Boolean {
    try {
      val signatureBase = Base64.decode(signature)
      return X25519Wrapper.verify(Base64.decode(publicKey), data.toByteArray(), signatureBase)
    } catch (e: Exception) {
      throw CryptoException(CryptoError.VerifyCryptoError, e)
    }
  }
}
